x-pack/plugins/proxy/server/proxy.ts

```diff
@@ -0,0 +1,276 @@
+      );
+      const res = h.response('Service Unavailable');
+      res.code(503);
+      return res;
+    }
+
+    url.hostname = node.node;
+    try {
+      const { payload } = await this.wreck.get(url.toString(), { json: true });
+      const res = h.response(payload);
+      res.code(200);
+      res.header('content-type', 'application/json');
+      return res;
+    } catch (err) {
+      this.log.warn(
+        `Unable to complete request to ${node.node} for ${resource} because ${err.message}`
+      );
+      const res = h.response('Internal server error');
+      res.code(500);
+      return res;
+    }
+  }
+
+  public getAllocation(): Observable<[string, RoutingNode]> {
+    return from(this.routingTable);
+  }
+
+  private setTimer() {
+    if (this.timer) return;
+    this.log.debug('Set timer to updateNodeMap');
+    this.timer = setInterval(async () => {
+      this.log.debug('Updating node map');
+      await this.mainLoop();
+    }, this.updateInterval);
+  }
+
+  private updateRoutingTable(routingTable: { [key: string]: RoutingNode }): void {
+    const currentRoutes = [...this.routingTable.keys()];
+    for (const [key, node] of Object.entries(routingTable)) {
+      this.routingTable.set(key, node);
+      const idx = currentRoutes.findIndex(k => k === key);
+      if (idx) currentRoutes.splice(idx, 1);
+    }
+
+    for (const key of currentRoutes.values()) {
+      this.routingTable.delete(key);
+    }
+  }
+
+  private async getNodeList(): Promise<NodeList> {
+    if (!this.elasticsearch) {
+      throw new Error('You must call setup first');
+    }
+    const client = await this.elasticsearch.pipe(first()).toPromise();
+    const params = {
+      id: this.proxyDoc,
+      index: this.proxyIndex,
+      _source: true,
+    };
+    const data: ClusterDoc = await client.callAsInternalUser('get', params);
+    this.updateRoutingTable(data.routing_table);
+    const nodes: NodeList = new Map(Object.entries(data.nodes));
+    return nodes;
+  }
+
+  private async updateNodeList(nodes: NodeList): Promise<void> {
+    if (!this.elasticsearch) {
+      throw new Error('You must call setup first');
+    }
+    const doc = {
+      nodes: map2obj(nodes),
+      routing_table: map2obj(this.routingTable),
+    };
+    const client = await this.elasticsearch.pipe(first()).toPromise();
+    const params = {
+      id: this.proxyDoc,
+      index: this.proxyIndex,
+      doc,
+    };
+    await client.callAsInternalUser('update', params);
+  }
+
+  private updateLocalNode(nodes: NodeList, finishTime: number): NodeList {
+    nodes.set(this.nodeName, {
+      lastCheck: finishTime,
+      lastUpdate: finishTime,
+    });
+    return nodes;
+  }
+
+  private async mainLoop(): Promise<any> {
+    const nodes = await this.getNodeList();
+    const finishTime = new Date().getTime();
+
+    for (const [key, node] of nodes.entries()) {
+      if (!node || finishTime - node.lastUpdate) {
```

Comment:
@toddself I found the code here to be different from what I thought.
Also, I found some errors in the [doc](https://docs.google.com/document/d/1xQGff7dS69G1lULD3QT2_DlBTLDFyFrDMtufQH9-pw0/edit#heading=h.91d5gfv7jlys) may cause this misunderstanding, so I fixed it.

In order to eliminate the problem caused by clock inconsistency in different nodes, the pseudo code in the document records the update time of each node in the memory of each kibana node. For the node that currently performs liveness check, the absolute value of node.lastUpdate is not important. The important thing is that 'I saw that node changed lastUpdate', so it must be online.

```
interface Liveness {
  lastUpdate: Date // Last time that a Kibana node (could be this or other nodes) update its liveness info in ES)
  lastChangedTime: Date  // Last time that current server found `lastUpdate` changed
}

const livenessMap: Map<UUID, Liveness> = {}  //  in local memory.

function updateLivenessMap(doc: any, currentTime: number) {
  // compare values in doc, if a value changes, update
  for (const uuid of nodes) {
    if (livenessMap[uuid].lastUpdate != doc.nodes[uuid].lastUpdated) {  // I see that node changed its lastUpdate
       livenessMap[uuid].lastCheck = currentTime;  // I record this changed acording to my clock time in my memory
       livenessMap[uuid].lastUpdate = doc.nodes[uuid].lastUpdate;
    } 
  }
}

// I found some nodes hasn't been change its `lastUpdate` for a while, so I think they are offline.
const timeoutNodes = livenessMap.values().filter(n => (finishTime - n.lastChangedTime) > timeoutThreshold)

```

Reply:
excellent thank you!

Reply:
Giving this some more thought.

If the idea is that we want to drop nodes that haven't updated the elasticsearch document "recently enough", this original logic should work and we don't have to take on the burden of tracking the nodes locally.

Looping through the response from elastic about what nodes are available and dropping the nodes that haven't updated longer than the `timeoutThreshold` should fix this issue even if there is clock skew between kibana instances. 

I just updated this to correctly account for the timeoutThreshold.  I suppose we should warn if the `timeoutThreshold` set in the preferences isn't larger than the `refreshInterval` as well